### ISSUE TITLE
FEP Module

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1931]
I propose that we add a new FEP (free energy perturbation) module, as `dc.fep`. 

Introduction
------------
Free energy perturbation has become an increasingly powerful technique in modern drug discovery. Starting with the publication of techniques like [LOMAP](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3837551/) and Schrodinger's FEP+ [technique](https://pubs.acs.org/doi/10.1021/ja512751q), and the release of open source tools like [Yank](http://getyank.org/latest/), free energy techniques have matured as powerful techniques for estimating the binding free energy of molecules to proteins. The basic idea of FEP is that is it's possible to estimate the binding free energy of small changes to a system by using Zwanzig FEP identity (Github doesn't support latex so this will be a little messy)

```
E_A[exp(-beta*Delta(U))] = exp(-beta Delta(F))
```

Here `U` is an energy function. We assume we have two states `A` and `B`. Think of `A` as the initial state, with energy function `U_A` and `B` as the ending state with energy function `U_B`. The difference `Delta(U) = U_B - U_A` The expectation `E_A` is over the distribution specified by the density

```
p_A = C * exp(-beta U_A)
```

In practice, it's possible to perform a molecular dynamics simulation to compute this expectation. This allows for the estimation of the change in free energy `Delta(F) = F_B - F_A` from a simulation in `A`. For this simulation to converge reasonably though, `B` and `A` should overlap considerable (since we are sampling from `A` to estimate the density of `B`). This has traditionally been sampling. There are techniques like MBAR (see https://github.com/choderalab/pymbar, https://github.com/alchemistry/alchemlyb) that help perform these calculations, but convergence can still be slow.

DeepMind recently came out with a fascinating paper [Targeted Free Energy Estimation Via Learned Mappings](https://arxiv.org/pdf/2002.04913.pdf) that proposes a technique to help with this problem. The idea is to use a *normalizing flow* that transforms state `A` to state `A'` that has higher overlap with `B`. A normalizing flow is a type of deep network that evolves probability distributions. The DeepMind paper trains a normalizing flow for a simple solute-solvent system. Training data is generated by a MD simulation and used to train the normalizing flow. Results on this simple system show that the use of then normalizing flow appears to considerably speed up the convergence of the system.

Proposed Changes
-------------------
I propose we should add support for deep FEP models in DeepChem. Doing this would require the following steps:

- Adding support for normalizing flows: Normalizing flows behave a little differently from our supervised/metalearning/RL models since they evolve distributions. This would require some new infrastructure. There are luckily a number of reference normalizing flow implementations out there already which have permissive licenses (https://github.com/tonyduan/normalizing-flows), so we could likely leverage this code to build out some infrastructure.
  - This code should probably go in `dc.models` although if models are different enough, we might need to make a `dc.normflow` submodule like we have for `dc.rl` and `dc.metalearning`.
- We need to add additional metrics suitable for normalizing flow training and evaluation to `dc.metrics`.
- We need to add new loss functions, the LBAR (Learned Bennett Acceptance Ratio) and LFEP (Learned Free Energy Perturbations) as defined in the DeepMind paper.
- To make deep FEP models more broadly useful, we need to generate larger and more interesting training datasets for them. These datasets should eventually find their way into `dc.molnet`. These datasets likely have to be generated by MD simulation of many protein/ligand systems. This will likely be a large effort and ideally we should find a way to tap already existing databases. Folding@Home has been running a lot of free energy calculations so they might have a suitable dataset already.
- To apply FEP on systems of practical interest, we need a number of additional tools. When applying FEP on lead optimization problems, usually a series of related compounds are constructed. I propose that we add utilities to do this automatically, using CReM (https://github.com/DrrDom/crem), a library to construct chemically reasonable mutations of a starting compound. The general pattern here is of a `PerturbationGenerator` abstract class that generates perturbed versions of an initial state. The CReM class would probably be a `MoleculePerturbationGenerator` concrete subclass. This should likely live in the `dc.fep` module.
- We need a way of running FEP on new systems of interest. One way to do this might be by adding a new `FEPEngine` class. Underneath the hood, this class should rely on Yank and OpenMM to the degree possible. This would live in the `dc.fep` module.

Scope
------
Would these changes be in scope for DeepChem? One possibility to ask is wouldn't this be a better contribution to Yank or a separate library? When eventually, as normalizing flow techniques mature, they will likely find their way into libraries like Yank that focus on free energy perturbation. But for the moment, normalizing flow techniques are very new. The DeepMind paper focuses on a very toy system. Considerable research and development will have to be done before these techniques are suitable for broader applications. This will involve a lot of model building, dataset gathering, benchmarking etc. As a scientific deep learning library, DeepChem is well suited to help facilitate these types of activities. I believe DeepMind has not open sourced their implementation, so creating a high quality reference implementation will help accelerate research in this field and get these techniques closer to practical applicability.

As a second question, it's reasonable to ask whether this should be it's own library instead of a part of DeepChem. The major advantage of building it within deepchem is that it's easy to leverage and extend the work we've put into build/documentation/tooling around DeepChem. Bootstrapping a new library would be considerable work for a very experimental technique.

Another thing to note is that I'm in the middle of overhauling DeepChem's support for structure based drug discovery at the moment. We have the new `dc.docking` module and I'm working on extending atomic convolutions. There should be natural synergy between these efforts and `dc.fep` that should be mutually beneficial. 

Implementation
---------------
I'm willing to take the lead on implementing this feature, but since this is a large set of new features, any help other folks are interested in providing would be much appreciated. Also, everything I've laid out in this issue is just a first design sketch. Feedback and comments are very welcome!

CC @peastman @ncfrey who I think would be interested.

**Comment by ncfrey** [comment]
@rbharath I think this sounds really exciting and I would be happy to help! I've been reading the DeepMind paper, and also reached out to the corresponding authors to see if they would be willing to share anything, but haven't heard back from them. 

I think it makes sense to incorporate normalizing flows into DeepChem. A reference implementation and tutorial would be really useful. Let me know if you'd like to discuss more details.

**Comment by peastman** [comment]
cc @proteneer and @jchodera who have experience in this area.  I need to read the DeepMind paper.  Then I can offer feedback.

**Comment by proteneer** [comment]
My initial 2 cents, while the paper is quite interesting from a theoretical point of view, I think there are lower hanging fruits to go after in the structure based drug discovery world before tackling FEP.

**Comment by peastman** [comment]
That's a really interesting paper.  I don't think we should add a `dc.fep` module though, for the same reason we don't have a `dc.toxicity` or `dc.solubility` module: as far as deep learning is concerned, FEP is an application, not a tool.  A very speculative, unproven application at that.

I do think it would be good to add most of the other features you suggest.  Let's provide the tools for people who want to do research on this, and perhaps some day it will mature to the point where it becomes a useful tool in itself.

**Comment by rbharath** [comment]
One thought here is that we should perhaps add a `dc.applications` submodule. `dc.docking` is already a bit of an odd-duck since unlike the other top-level submodules, it's not a generic ML tool. `dc.docking` might better fit as `dc.applications.docking`. We could similarly add any needed FEP-specific tooling as `dc.applications.fep`.

**Comment by ncfrey** [comment]
Based on the DeepMind paper, this [review article](https://arxiv.org/abs/1912.02762) and this paper on [normalizing flows for molecular graph generation](https://arxiv.org/pdf/1905.11600.pdf), I've tried to come up with some building blocks that are within the scope of DeepChem and should be useful for anyone who wants to use normalizing flows.

- `NormalizingFlows` superclass to specify the API (suggested by @rbharath)
- Coupling layers
- Circular spline bijectors, which are the same as tfp.bijectors.RationalQuadraticSpline, except compatible with periodic boundary conditions
- Kullback-Leibler divergence
- Transformers with multi-headed self-attention layers

These features would be generally useful for generative modeling and FEP.

**Comment by rbharath** [comment]
We had FEP come up in the DeepChem office hours discussion. Now that we have normalizing flows in DeepChem, do we have the infrastructure needed to implement the DeepMind paper in place?

CC @ncfrey

**Comment by ncfrey** [comment]
I think the only thing we need would be a custom implementation of a circular spline bijector, which could either inherit from ` tfp.bijectors.RationalQuadraticSpline` or use its code as a starting point and be modified to respect PBC.

**Comment by rbharath** [comment]
One thought here is that we might want to implement a `dc.dock` like module for programmatic FEP (calling say Yank or a new normalizing flow engine) under the hood. I don't yet have a sense of how complicated this would be though
